python/dglke/tests/test_infer.py

```diff
@@ -0,0 +1,270 @@
+    tail_emb = F.uniform((num_tail, dim), F.float32, F.cpu(), 0, 1)
+    rel_emb = F.uniform((num_rels, dim), F.float32, F.cpu(), -1, 1)
+
+    if func_name == 'RotatE':
+        rel_emb = F.uniform((num_rels, dim//2), F.float32, F.cpu(), -1, 1)
+    if func_name == 'RESCAL':
+        rel_emb = F.uniform((num_rels, dim * dim), F.float32, F.cpu(), -1, 1)
+
+    if func_name == 'TransE':
+        return head_emb, rel_emb, tail_emb, (12.0)
+    elif func_name == 'TransE_l1':
+        return head_emb, rel_emb, tail_emb, (12.0, 'l1')
+    elif func_name == 'TransE_l2':
+        return head_emb, rel_emb, tail_emb, (12.0, 'l2')
+    elif func_name == 'RESCAL':
+        return head_emb, rel_emb, tail_emb, (dim, dim)
+    elif func_name == 'RotatE':
+        return head_emb, rel_emb, tail_emb, (12.0, 1.0)
+    else:
+        return head_emb, rel_emb, tail_emb, None
+
+ke_infer_funcs = {'TransE': TransEScore,
+                  'TransE_l1': TransEScore,
+                  'TransE_l2': TransEScore,
+                  'DistMult': DistMultScore,
+                  'ComplEx': ComplExScore,
+                  'RESCAL': RESCALScore,
+                  'TransR': TransRScore,
+                  'RotatE': RotatEScore}
+
+class FakeEdge:
+    def __init__(self, hemb, temb, remb):
+        data = {}
+        data['head_emb'] = hemb
+        data['tail_emb'] = temb
+        data['emb'] = remb
+        src = {}
+        src['emb'] = hemb
+        dst = {}
+        dst['emb'] = temb
+
+        self._src = src
+        self._dst = dst
+        self._data = data
+
+    @property
+    def src(self):
+        return self._src
+
+    @property
+    def dst(self):
+        return self._dst
+
+    @property
+    def data(self):
+        return self._data
+
+class InferKEModel:
+    def __init__(self, score_func, entity_emb, rel_emb):
+        self.score_func = score_func
+        self.entity_emb = entity_emb
+        self.rel_emb = rel_emb
+
+    def infer_score(self):
+        head_emb = generate_rand_emb
+
+def check_infer_score(func_name):
+    batch_size = 10
+
+    ke_score_func = ke_infer_funcs[func_name]
+
+    # normal
+    head_emb, rel_emb, tail_emb, args = generate_rand_emb(func_name, 'none')
+    if args is None:
+        score_func = ke_score_func()
+    elif type(args) is tuple:
+        score_func = ke_score_func(*list(args))
+    else:
+        score_func = ke_score_func(args)
+    score1 = score_func.infer(head_emb, rel_emb, tail_emb)
+    assert(score1.shape[0] == head_emb.shape[0])
+    h_score = []
+    for i in range(head_emb.shape[0]):
+        r_score = []
+        for j in range(rel_emb.shape[0]):
+            t_score = []
+            for k in range(tail_emb.shape[0]):
+                hemb = head_emb[i]
+                remb = rel_emb[j]
+                temb = F.unsqueeze(tail_emb[k], dim=0)
+                edge = FakeEdge(hemb, temb, remb)
+                score = score_func.edge_func(edge)['score']
```

Comment:
the difference between `infer` and `edge_func` is that `infer` computes scores in a batch fashion?

Reply:
infer will do broadcasting as (N, 1) * (1, N).
edge_func only do N * N.